src/main/java/edu/ucla/library/bucketeer/verticles/FinalizeJobVerticle.java

```diff
@@ -170,26 +282,158 @@ private void failPromise(final Throwable aException, final String aMessageCode,
      * @param aChannelId ID of the channel to which we want to send this message
      */
     private void sendSlackMessage(final String aChannelId, final String aMessageText) {
-        sendSlackMessage(aChannelId, aMessageText, null);
+        sendSlackMessage(aChannelId, aMessageText, null, null);
     }
 
     /**
      * Send a message to the specified Slack channel with a list of metadata records.
      *
      * @param aMessageText Text of the message we want to send
      * @param aChannelId ID of the channel to which we want to send this message
+     * @param aJob A job to notify a Slack user of the completion of
+     * @param aCsvData The CSV data to send as a file attachment in the Slack message
      */
-    private void sendSlackMessage(final String aChannelId, final String aMessageText, final Job aJob) {
+    private void sendSlackMessage(final String aChannelId, final String aMessageText, final Job aJob,
+            final String aCsvData) {
         final JsonObject message = new JsonObject();
 
         message.put(Config.SLACK_CHANNEL_ID, aChannelId);
         message.put(Constants.SLACK_MESSAGE_TEXT, aMessageText);
 
-        if (aJob != null) {
+        if (aJob != null && aCsvData != null) {
             message.put(Constants.JOB_NAME, aJob.getName());
-            message.put(Constants.BATCH_METADATA, JsonObject.mapFrom(aJob));
+            message.put(Constants.CSV_DATA, aCsvData);
         }
 
         sendMessage(message, SlackMessageVerticle.class.getName());
     }
+
+    /**
+     * Update the job's metadata before it's turned back into CSV and output to Slack.
+     *
+     * @param aJob A job whose metadata we want to update
+     * @return The job with the updated metadata
+     */
+    private Job updateMetadata(final Job aJob) throws ProcessingException {
+        final List<String[]> metadata = aJob.getMetadata();
+        final List<Item> items = aJob.getItems();
+
+        String[] metadataHeader = aJob.getMetadataHeader();
+        int bucketeerStateIndex = -1;
+        int accessUrlIndex = -1;
+
+        // Find the index position of our two columns: Bucketeer State and Access URL
+        for (int headerIndex = 0; headerIndex < metadataHeader.length; headerIndex++) {
+            if (Metadata.IIIF_ACCESS_URL.equals(metadataHeader[headerIndex])) {
+                LOGGER.debug(MessageCodes.BUCKETEER_154, headerIndex);
+                accessUrlIndex = headerIndex;
+            } else if (Metadata.BUCKETEER_STATE.equals(metadataHeader[headerIndex])) {
+                LOGGER.debug(MessageCodes.BUCKETEER_153, headerIndex);
+                bucketeerStateIndex = headerIndex;
+            }
+        }
+
+        // If both headers are missing, we need to expand our headers array by two
+        if (bucketeerStateIndex == -1 && accessUrlIndex == -1) {
```

Comment:
I wonder if it would be better to define a NOT_FOUND (or something) constant for -1 since we use that in multiple places? I'm fine with either way, just a thought/suggestion.

Reply:
Or we could store the value of each of those Boolean expressions in Boolean variables? That would be more readable IMO.

Reply:
I hadn't really looked at this method since I just moved it from one class to another, but now that I'm looking at it again, I wonder if it should go into the Job class instead (since it's returning the same Job object that gets passed to it).

Reply:
Yes for booleans where the actual index isn't needed. I think in some places we're using the index positions to pull contents from an array though?

Reply:
Interesting point about Job. I think I'd thought of that as just a container interested in the mechanics of the job processing, but I don't have any objections to it updating metadata. It does take and return a job as you say, which is a bit odd.